### ISSUE TITLE
[BUG] Unable to add node to RK2 cluster using node driver

### ISSUE THREAD

**Issue by abonillabeeche** [harvester/harvester#1823]
**Describe the bug**
After creating an RKE2 cluster, I am unable to add an additional node to the cluster by clicking +

**To Reproduce**
1. Create a cluster, with a single node pool
2. Click + to add another node to the node pool
3. Wait - see as the node is created in Harvester but never acknowledged by Rancher

**Expected behavior**
Another node should be added to the cluster

**Support bundle**
https://drive.google.com/file/d/16bgxeR5YeB2boqywmInSc-paxqMx-onG/view?usp=sharing

**Environment:**
 - Harvester ISO version: 1.0
 - Underlying Infrastructure: R620

**Additional context**
This should be easily reproducible

**Comment by FrankYang0529** [comment]
I can reproduce this issue with `SLES15-SP3-JeOS.x86_64-15.3-OpenStack-Cloud-GM.qcow2`, but not `focal-server-cloudimg-amd64.img`. Following is the error message in `rancher/machine` container.

```
Error creating machine: Error saving host to store after attempting creation: unable to load secret : error getting secret from kubernetes: Unauthorized
The default lines below are for a sh/bash shell, you can specify the shell you're using, with the --shell flag.
```

**Comment by abonillabeeche** [comment]
Correct. I can't reproduce with focal-server-cloudimg-amd64.img. Yet, while testing this, I clicked the - sign to delete that node I had just recently added, and my Harvester and Rancher clusters became unresponsive. @FrankYang0529 can you try and see if that happens to you too?

**Comment by FrankYang0529** [comment]
Looks like the issue isn't always reproducible. I can't reproduce with the same steps in a new Harvester cluster and Rancher.

In my old cluster, the error message is not always like my previous comment. Sometimes, the Job in Rancher is killed before the VM is ready, so the status of the Machine in Rancher is stuck in Provisioning. I have no idea which process kills the Job yet.

I think there may be a timeout config somewhere. If the Job is running for a long time and it will be killed. In my cluster, SLES takes 6m to finish `rancher/machine` Job. Focal only takes around 2m.

![Screen Shot 2022-01-19 at 2 23 40 PM](https://user-images.githubusercontent.com/4927361/150076608-13a25d62-2b98-4dd6-9e42-3b2fed79e411.png)
![Screen Shot 2022-01-19 at 2 29 09 PM](https://user-images.githubusercontent.com/4927361/150076621-334c336d-9716-4c3d-b8b8-1eace297ca74.png)

**Comment by FrankYang0529** [comment]
> Correct. I can't reproduce with focal-server-cloudimg-amd64.img. Yet, while testing this, I clicked the - sign to delete that node I had just recently added, and my Harvester and Rancher clusters became unresponsive. @FrankYang0529 can you try and see if that happens to you too?

@abonillabeeche, what do you mean about unresponsive? If a Machine in Rancher loses connection to a Harvester VM, the Machine CR can't be deleted caused of finalizers. My cluster still works, but I have to manually remove finalizers to remove Machine CR.

**Comment by FrankYang0529** [comment]
I can reproduce this issue sometime. If a job is deleted occasionally, we will get log like the following. The VM information is not stored to the Machine in Rancher, so the status of the Machine is stuck in Provisioning.
![Screen Shot 2022-01-20 at 5 46 35 PM](https://user-images.githubusercontent.com/4927361/150314616-4ecfb5aa-8a51-4558-a2ad-9ca4c7aa33e2.png)

Also, the job which doesn't really run for a long time still can be removed. From my previous comment, a 6m job can finish, but a 4m job may be killed.
![Screen Shot 2022-01-20 at 6 05 22 PM](https://user-images.githubusercontent.com/4927361/150318027-311f110c-b558-4774-a873-5efeff18acf7.png)


If a job is finished successfully, we can get log like the following.
![Screen Shot 2022-01-19 at 2 11 09 PM](https://user-images.githubusercontent.com/4927361/150314744-99e03acf-b42c-4c6a-9f5d-29dba301992d.png)

I still can't find which controller removes the job, but we can enable [Job tracking with finalizers](https://kubernetes.io/docs/concepts/workloads/controllers/job/#job-tracking-with-finalizers) after k8s 1.23.

**Comment by a-blender** [comment]
This is a bug in Rancher. As Rancher engineers, we cannot transfer this issue to the `rancher/rancher` repo because of lack of permissions. Could you please transfer it for us?

**Comment by thedadams** [comment]
@FrankYang0529 Before we confirm that this is a bug in Rancher, I would like to clarify a few things.

1. Is the job being killed or deleted? There is a big difference here. There is nothing in Rancher that will stop the job. There is something that will delete the job, but only after certain conditions are met on the Harvester machine object. I don't see any of the machine objects in the dump for this issue, so I can't be sure that these condition were or were not met for Rancher to clean up the job.
2. What is the exit status of the pod? The way rancher-machine works is the node drivers are started as gRPC servers and rancher-machine will send requests to them. The logs for the node drivers, therefore, don't show up in the logs for the pod because those are rancher-machine's logs. If the gRPC server (in this case, the Harvester node driver) exited uncleanly, then the pod would stop unexpectedly. In order to determine if this is the case, the pod would have to be inspected more closely. Depending on how far rancher-machine got, I would not expect the job to be deleted by the Rancher controllers.

**Comment by FrankYang0529** [comment]
@thedadams Because I can't get logs or YAML from a deleted pod/job, I try to do some experiments.

* Delete Job: I can get a similar result like my previous comment. Job and pod are all removed.
![Screen Shot 2022-02-14 at 9 56 43 PM](https://user-images.githubusercontent.com/4927361/153877489-883a5b25-38d0-4984-ba73-b685b85da446.png)

* Delete Pod: I assume if the gRPC server has an error and pod will be terminated, but the job will not. In this case, I can see another pod is created and has logs like the following. Also, the new pod will be in `Error` status.
![Screen Shot 2022-02-14 at 9 46 49 PM](https://user-images.githubusercontent.com/4927361/153876910-2aa93926-5e30-465a-b055-0b8d79a6b60d.png)

**Comment by thedadams** [comment]
@FrankYang0529 Unfortunately, deleting the pod or job manually doesn't give enough information into why this is happening. In order to fix the problem, we need to understand why the job or pod is being deleted by Rancher. It is very logical that deleting the job or pod manually will produce similar results to when it is deleted by Rancher, but we need to figure out what is going on.

The way I think is best to do this is to the following two commands while the job is running:
- `kubectl -n fleet-default get jobs -o yaml -w`
- `kubectl -n fleet-default get pods -o yaml -w`

These commands will output the information needed as the job is executing. If the job is deleted or fails, then the output of those two commands will gives us a clue as to how this is happening.

**Comment by FrankYang0529** [comment]
@thedadams thanks for your help. I can get the container exit code now! It's `137`. Looks like it's `OOMKilled`.

```
apiVersion: v1
kind: Pod
...
status:
  conditions:
  - lastProbeTime: null
    lastTransitionTime: "2022-02-15T05:15:35Z"
    status: "True"
    type: Initialized
  - lastProbeTime: null
    lastTransitionTime: "2022-02-15T05:20:40Z"
    message: 'containers with unready status: [machine]'
    reason: ContainersNotReady
    status: "False"
    type: Ready
  - lastProbeTime: null
    lastTransitionTime: "2022-02-15T05:20:40Z"
    message: 'containers with unready status: [machine]'
    reason: ContainersNotReady
    status: "False"
    type: ContainersReady
  - lastProbeTime: null
    lastTransitionTime: "2022-02-15T05:15:35Z"
    status: "True"
    type: PodScheduled
  containerStatuses:
  - containerID: containerd://fded6a763f5289f4abd95cebe536414bb188fed5e7fd7efb9be6c8f4edfa6946
    image: docker.io/rancher/machine:v0.15.0-rancher73
    imageID: docker.io/rancher/machine@sha256:3baf7cb8bbc29fe9c16d45cd2b6c1a08a4deeb1bf6caff656fa264a4110c0d74
    lastState: {}
    name: machine
    ready: false
    restartCount: 0
    started: false
    state:
      terminated:
        containerID: containerd://fded6a763f5289f4abd95cebe536414bb188fed5e7fd7efb9be6c8f4edfa6946
        exitCode: 137
        finishedAt: "2022-02-15T05:20:39Z"
        message: |
          Downloading driver from https://releases.rancher.com/harvester-node-driver/v0.3.4/docker-machine-driver-harvester-amd64.tar.gz
          docker-machine-driver-harvester-amd64.tar.gz
          docker-machine-driver-harvester-amd64.tar.gz: gzip compressed data, from Unix, original size 36115968
          Running pre-create checks...
          Creating machine...
          (rke2-in-jeos-4-pool1-8ace122d-vkmvc) Creating New SSH Key
          (rke2-in-jeos-4-pool1-8ace122d-vkmvc) Using public Key: /home/machine/.docker/machine/machines/rke2-in-jeos-4-pool1-8ace122d-vkmvc/id_rsa.pub
          (rke2-in-jeos-4-pool1-8ace122d-vkmvc) Start node
          (rke2-in-jeos-4-pool1-8ace122d-vkmvc) Waiting for node become Running
          (rke2-in-jeos-4-pool1-8ace122d-vkmvc) Waiting for node get ip
        reason: Error
        startedAt: "2022-02-15T05:15:38Z"
  hostIP: 172.17.0.2
  phase: Failed
  podIP: 10.42.0.24
  podIPs:
  - ip: 10.42.0.24
  qosClass: BestEffort
  startTime: "2022-02-15T05:15:35Z"
```

**Comment by FrankYang0529** [comment]
@thedadams , I run rancher as a docker container with `--oom-kill-disable`. I can get logs like the following in the job.

```
Error creating machine: Error saving host to store after attempting creation: unable to load secret : error getting secret from kubernetes: Unauthorized
The default lines below are for a sh/bash shell, you can specify the shell you're using, with the --shell flag.
```

In `kubectl -n fleet-default get pods -o yaml -w`, I also get exit code 137. I think they are related. From source code, above logs are from [here](https://github.com/rancher/machine/blob/7766c7065bfb384292c51747010c1e165ce6481e/libmachine/persist/secretstore.go#L88-L91). Do you know what situation may cause OOM kill here? Thank you.

```
apiVersion: v1
kind: Pod
...
status:
  conditions:
  - lastProbeTime: null
    lastTransitionTime: "2022-02-15T08:13:44Z"
    status: "True"
    type: Initialized
  - lastProbeTime: null
    lastTransitionTime: "2022-02-15T08:20:06Z"
    message: 'containers with unready status: [machine]'
    reason: ContainersNotReady
    status: "False"
    type: Ready
  - lastProbeTime: null
    lastTransitionTime: "2022-02-15T08:20:06Z"
    message: 'containers with unready status: [machine]'
    reason: ContainersNotReady
    status: "False"
    type: ContainersReady
  - lastProbeTime: null
    lastTransitionTime: "2022-02-15T08:13:44Z"
    status: "True"
    type: PodScheduled
  containerStatuses:
  - image: rancher/machine:v0.15.0-rancher73
    imageID: ""
    lastState:
      terminated:
        exitCode: 137
        finishedAt: null
        message: The container could not be located when the pod was deleted.  The
          container used to be Running
        reason: ContainerStatusUnknown
        startedAt: null
    name: machine
    ready: false
    restartCount: 0
    started: false
    state:
      waiting:
        reason: ContainerCreating
  hostIP: 172.17.0.2
  phase: Failed
  podIP: 10.42.0.13
  podIPs:
  - ip: 10.42.0.13
  qosClass: BestEffort
  startTime: "2022-02-15T08:13:44Z"
```

**Comment by thedadams** [comment]
@FrankYang0529 Thanks for this info. I have never seen the machine create pod get OOM killed in any other situation. This may be an issue with the harvester docker-machine driver. It will take some time to find the problem.

**Comment by thedadams** [comment]
@FrankYang0529 Very interesting. I have never seen the machine create pod get OOM-killed. In order to investigate the problem with the `Unauthorized` error, I would need the job YAML to see the arguments, service account and secrets passed to the job.

It would also be helpful to see the `harvester.rke-machine.cattle.io` that was used to create the machine.

**Comment by FrankYang0529** [comment]
> @FrankYang0529 Very interesting. I have never seen the machine create pod get OOM-killed. In order to investigate the problem with the `Unauthorized` error, I would need the job YAML to see the arguments, service account and secrets passed to the job.
> 
> It would also be helpful to see the `harvester.rke-machine.cattle.io` that was used to create the machine.


* Job
```
apiVersion: v1
items:
- apiVersion: batch/v1
  kind: Job
  metadata:
    annotations:
      objectset.rio.cattle.io/applied: H4sIAAAAAAAA/6RV247bNhD9lYLPom62ZVuAHjYBkrRpLhv0skmRB4ocWYx5UUlKjh343wtK8iXe9WaDvtmc4eGZozMz35AERxhxBOXfEFFKO+K4Vtb/1eUXoM6CCw3XISXOCQi5jjhDOULB1bjeKDB41a1RjswasCS05grOMrok+OU1V6x4RUwH1oF5M+T8EFURCQNsirnCX0Bb3GgtEjzJlouYLUpM56zcPAnINoR6tEoAOMygIq1waB8gaqCX4Q8uwToiG5SrVogACVKCeFScmtga5SjO0oSSMkvm2XQ5IymwpITlkqbxdF6VSRZX2SIlceVfe2pNByFxY3THLdcKDZev1mEboJ5sSehaV9XvXHKH8jhADmQjiAMfPHfAEwo36/MvyVVlyJHYyui2ufbVUfDo3TVX3lcPOOKxWz/hh8dgOjC9njnqkiupBqTuepWJsID2Z/JSrRzhCoxF+T/fEDEr/wNhzAzvwGCmN0pownBrRFE719g8igwIIBZsaIiiNZiQahnVh+qx0gzG+1EXh5NwGjFN12COnEfw0xUiWTYNHTHhaoeC0/PekgWkyZTOSjJZlItJBSWFdLbIJmkyqxaLeTpZkgpgWU6n8zkh87icVvMsW87YbDKZ+7we0AI14E69U3zvuIuU4umOts6bcWw8OCNfHMvrD2lrnZaYK+uIENhSwxtXRKZV0fCwjUbIqCOiHZBOCtGmxVS3yhXpRYRxu8Zla4uOG9d79V7U8h0U0/gixCVZDTYsRhmi4Uh2cvnvRbIEqc12RLqIKXAbbdZYagbiYRqHlO9e6wRRyZVEt22gYDVtLuLW1rjRxhXppQ4+0lowhRVgL2L+HPtJUdCXL3Z0wjoqb7NfX30QdHL7+mNabz7evd2x9DZ7zm9Wn+7qDbn7Tbzjz+THv7/uPt02O5rWFX3Fanb3Qb/jz2Kq/hIXj3TyzF4nY/24wT8HCFT3wmjZN+Fghw9Q+Qb96Qk7ds4Agvb7zwHqv6oHGdr14LO8i8NkFsZ4PJ9P0Jj7vhXivRacblGObsSGbC06Dnt5mm5gdWso+OHqZwrQ1nC3fa6Vg6+un7iturEvh9GaxHEcDCd/WjDDwT5AnRathDfe28MQkv7ne+L8MnqoP05USq2ddYY0aB9c3ANHI2tFRME4G1GCCWPc7wciwgbkCcMJi78L44FQXx9h75TYotyZFgJk23KEv4+3/+yVNn71GHfU7i10fftbMB2ncEP7Fn77f7bmwG5Q6p4MwWger/3owDeaAcpnSRIg3QyEUd4vgkPyQ3SOkPiwb/FqM63nR0pn0nvtfyzmFWLpMj0nNgp9zusepjf1vl9ijri2N9/+vwAAAP//TCEVQQ4KAAA
      objectset.rio.cattle.io/id: ""
      objectset.rio.cattle.io/owner-gvk: rke-machine.cattle.io/v1, Kind=HarvesterMachine
      objectset.rio.cattle.io/owner-name: rke2-in-jeos-pool1-36980d8b-c7dbw
      objectset.rio.cattle.io/owner-namespace: fleet-default
    creationTimestamp: "2022-02-16T06:45:12Z"
    labels:
      objectset.rio.cattle.io/hash: 0621cab6176495a2ed1be99c2047fb160f682a0f
    name: rke2-in-jeos-pool1-36980d8b-c7dbw-machine-provision
    namespace: fleet-default
    resourceVersion: "72238"
    uid: 943c0813-331f-4186-a7fa-47aa3b65f84e
  spec:
    backoffLimit: 0
    completions: 1
    parallelism: 1
    selector:
      matchLabels:
        controller-uid: 943c0813-331f-4186-a7fa-47aa3b65f84e
    template:
      metadata:
        creationTimestamp: null
        labels:
          controller-uid: 943c0813-331f-4186-a7fa-47aa3b65f84e
          job-name: rke2-in-jeos-pool1-36980d8b-c7dbw-machine-provision
          rke.cattle.io/infra-machine-group: rke-machine.cattle.io
          rke.cattle.io/infra-machine-kind: HarvesterMachine
          rke.cattle.io/infra-machine-name: rke2-in-jeos-pool1-36980d8b-c7dbw
          rke.cattle.io/infra-machine-version: v1
          rke.cattle.io/infra-remove: "false"
      spec:
        containers:
        - args:
          - --driver-download-url=https://releases.rancher.com/harvester-node-driver/v0.3.4/docker-machine-driver-harvester-amd64.tar.gz
          - --driver-hash=e214c5ba38b83febce25863215f887239afee9b4477aa70b4f76695d53378632
          - --secret-namespace=fleet-default
          - --secret-name=rke2-in-jeos-pool1-36980d8b-c7dbw-machine-state
          - create
          - --driver=harvester
          - --custom-install-script=/run/secrets/machine/value
          - --harvester-cpu-count=2
          - --harvester-disk-bus=virtio
          - --harvester-disk-size=40
          - --harvester-image-name=default/image-mvm9q
          - --harvester-memory-size=4
          - --harvester-network-model=virtio
          - --harvester-network-name=default/vlan1
          - --harvester-network-type=dhcp
          - --harvester-ssh-port=22
          - --harvester-ssh-user=sles
          - --harvester-user-data=cGFzc3dvcmQ6IHRlc3QKY2hwYXNzd2Q6CiAgZXhwaXJlOiBmYWxzZQpzc2hfcHdhdXRoOiB0cnVl
          - --harvester-vm-namespace=default
          - rke2-in-jeos-pool1-36980d8b-c7dbw
          envFrom:
          - secretRef:
              name: rke2-in-jeos-pool1-36980d8b-c7dbw-machine-driver-secret
          image: rancher/machine:v0.15.0-rancher73
          imagePullPolicy: Always
          name: machine
          resources: {}
          securityContext:
            runAsGroup: 1000
            runAsUser: 1000
          terminationMessagePath: /dev/termination-log
          terminationMessagePolicy: File
          volumeMounts:
          - mountPath: /run/secrets/machine
            name: bootstrap
          - mountPath: /etc/ssl/certs/ca-additional.pem
            name: tls-ca-additional-volume
            readOnly: true
            subPath: ca-additional.pem
        dnsPolicy: ClusterFirst
        restartPolicy: Never
        schedulerName: default-scheduler
        securityContext: {}
        serviceAccount: rke2-in-jeos-pool1-36980d8b-c7dbw-machine-provision
        serviceAccountName: rke2-in-jeos-pool1-36980d8b-c7dbw-machine-provision
        terminationGracePeriodSeconds: 30
        volumes:
        - name: bootstrap
          secret:
            defaultMode: 511
            optional: false
            secretName: rke2-in-jeos-bootstrap-template-gw4h7-machine-bootstrap
        - name: tls-ca-additional-volume
          secret:
            defaultMode: 292
            optional: true
            secretName: tls-ca-additional
  status:
    active: 1
    startTime: "2022-02-16T06:45:12Z"
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""
```
* Role
```
apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
  annotations:
    objectset.rio.cattle.io/applied: H4sIAAAAAAAA/4yQMU/DMBCF/wq6OS5JKGkTiRkkBBMb6nC2L61bx7bsSzpU/e/IqZC6FDreu3vf6b0TDMSokRG6E6BznpGNdymPXu5JcSJeROMXCpktLYx/NBo6gOLm3h8dRbGdDtBBPJAYUO2Mo6uLqSoe3o3TL28YJ0pM8eNy8y/V4UAXbC2ME3vySQTvbSWemnZd6rUUaqXl8S5QCqgyrbdELDT1OFqGcwEq0lzDlxkoMQ4BOjdaW4BFSfbPcnaYdtBB2dSVQtlUq2bZPmNNupLUtqoul6teVk3ZN+sayz5/uzfTb5EiRD+ZZLyDi/lmjjhaStB9nwCDeY1+DHkC2BQQKfkxKvrM/ize/z4xMl0zZn8iFYlT1ieKcta2xFDAGPRsOG/OPwEAAP//fkS4m3ACAAA
    objectset.rio.cattle.io/id: ""
    objectset.rio.cattle.io/owner-gvk: rke-machine.cattle.io/v1, Kind=HarvesterMachine
    objectset.rio.cattle.io/owner-name: rke2-in-jeos-pool1-36980d8b-c7dbw
    objectset.rio.cattle.io/owner-namespace: fleet-default
  creationTimestamp: "2022-02-16T06:45:12Z"
  labels:
    objectset.rio.cattle.io/hash: 0621cab6176495a2ed1be99c2047fb160f682a0f
  name: rke2-in-jeos-pool1-36980d8b-c7dbw-machine-provision
  namespace: fleet-default
  resourceVersion: "73550"
  uid: d593957b-4c03-4f43-8bc4-d12d11e51d2d
rules:
- apiGroups:
  - ""
  resourceNames:
  - rke2-in-jeos-pool1-36980d8b-c7dbw-machine-state
  resources:
  - secrets
  verbs:
  - get
  - update
```
* Secret
```
apiVersion: v1
data:
  extractedConfig: H4sIAAAAAAAA/+xW22/6OBbu8/wVEc9DcRIuBWkeyiVAWlJysaEIaWVitwlxLrVNKF3N/75KgHL5tTOr1WhWq/2dPqT4O/5yzrFzvlML0pjWYuwHYUJrtyT1I8o/fx+eosYjqlXDpLqmqahmacrUqt5s3wFyt6r6LbLa3vyBAQBAC4DiqbYaF8+j3ah1Deiq2my0mjdAazbqjRul8Uekf5VthMT8BoBDqt/6/Rl+TOT4/B+xv+r8a36avISvt2uRJtfvKOrR/P78G3Xt6vx10Gq2bpS/pYj/5+f/z18URVEqvfL0EOUiTJNKR9F/3a/3eZhTXukoe79ybTy9J4RTISodpVL59QRM9iWycEwL6E/vzPle1x1BUb6pIhgVV9A05bLSUTTtcvmB7qZYBtdxuDLl9Aj84QW/2LXFPJ5gIcsoXjAT9BodpaIIoyL9rFOrgdvyr6PrreYPRP1Q+GlO+e46uIfNiu6r3UsTSZOSEGfhofYdJVeXSRQmpKPs3ZaJzzZFVKKzTKpKgmPaUZaVAPN8WVkminKAO8X/iiIozykvPaTMRKdWU9vardq8u1Xvbpu16E7Ujnw1vxpX2VuWk48123Mpik+5DF9CH0taxRsZpDyUuyrBEhecjy6Qj67acyAxvbDbd5AJXbgwbWBAp8SA7EWq6UJzSyOrby+Xe1pFQXC3AIbuQrO7AKxnQ6frAjNZ6EaKtWwKDcNxBsab7RELJkR3NLJwjIUDoTTooMEXyFfx0JyvhunD84k1IoLGzgglCM8GDDyvrYdVpLq+TkYQsQFiBSuqu7Fk6AOlz4wIHJuj1SiLHHWhzlhgEiPaOXMyxifWofXkzTKTbDMLR3DnQcekg/bA84w5LrB5ZhKgvjle13IGrO4Bx5og8nXsJ9aLLLLQ0d6bvf809hPrZRZzNDCgHZO1Pws4ju4eoGGYdnRRZ8uOTN2LrK6NzAaBVnMSmWAyUN8nJ9aZNXGH2x3RLfgMJx8LEBiuFkwnyEzpOhiT6Jk7/cBY2enHZH2v0qjxvNKNqadnJhwtEDEcRtcIIrjwZmeswdgxnM2KsZn34RjeoI1naJxD7W7rzklx/lM7Jk8oCpyngdldrAOjF5ldD1pdHxm6Myrr6k4MpwtPrPA1tyMDEWDoDhznNDKiI4s1QMOFaqBnhB5nM5GTnQztSAJHXWBLD8aumj74fcsmavA4m5P+4qyuzHS1Nnbn2dtkgLoEsAEERpeopuPOgq4bBQmE7dxXbc0a3u1s3a/bLHgnjOmOJofEQGK2Dp5sO5Nnd2Bk7iwvsFfzrmuvWRdDo/k8M5hnMN2aBWMLDBqob9qzdXeBoBVAhB5hks2toV9fzO8BnhujVRzE0Ojq0xOrBx4OX6bhRfarDZCLBmzoQquLBrD8Kh/t334rvvFlshHf95ICOzQSmUY0KdBos6J7da8WcLW1ZR/RGwlYs9PQslYrF6ucbXP52mw3tfiNZI3mm/ba3DZaq/XLW53xl62I/Zz5zbjOG0Fruw9jmfhFA3yX3/a1PXwIp4zsAj91vrPlZeJvOKeJrB63f4LnXRhNCpUSGfZLqSL0BW+YPPfo7bm9XVZ6hHGWcknJFy7jfuFw3ksv1KDsxydlrcSlxFTboK2SFgHVgOJL1ikspO5cV2mc8p0bfpSR1IfhuXs/FNEnBH7EuptSp/OQyzA9B8cxfv3U6kMBamGxWI3zuP12IVqF0Ib86H2tzjzMsaTfqbE7mm5WLPQf6A9iWGBYiG3KyTV0T0gxEHhpv1TsIU832Re6bFG5TXl0PCYS+FnlR/g6zZzhRP3Cb5ISyr4uV4+lG7JX5HKskHFWi1MSvoSU7L+LQiN1DdQbjbuWfr61yKOPJS72ZYd0O4qkQi4TPyhXSHnL6XsWcnpIcpkIEfwj2xYi3FEk39AvAj7SVkrk94up7Zh0cftpOdYc4GKMecpkmCbicrL7nPYux7ry+lU6Cri6WldLvY2QaTxOhMSMuT4Ps3K2qfFNUhPU51SKzzksx+wyn0HyGib0FFWyYex6pvoWvd/I4BK8KMe/P5D+8vt/ex7/aT/tp/20v8v+FQAA///JDYqqABQAAA==
kind: Secret
metadata:
  annotations:
    objectset.rio.cattle.io/applied: H4sIAAAAAAAA/4yQz07DMAzGXwX53JSmjPSPxB0JceMFnMSl2dKkSrxOaNq7ozKQxoGxo+VPP/v7HWEiRouM0B8BQ4iM7GLI6xj1lgxn4jK5WBpk9lS6eO8s9ADFn/t4CJTE+7KDHtKOxIRmdIEuEoss7l5csE/PmBbKTOn1nPmXGnCiM7YWLogtxSzmGL0UD6prK9tqYRqrDzeB8oxmpQ2eiIWlAfee4VSASfSl4c1NlBmnGfqw974Aj5r8VTkj5hF6qFQtDWolG7XpHrEmKzV1namrTTNoqapBtTVWw3rt1k4/IkVm5FXV1Q78MX9TL777TTh9BgAA///Uu5t8/wEAAA
    objectset.rio.cattle.io/id: ""
    objectset.rio.cattle.io/owner-gvk: rke-machine.cattle.io/v1, Kind=HarvesterMachine
    objectset.rio.cattle.io/owner-name: rke2-in-jeos-pool1-36980d8b-c7dbw
    objectset.rio.cattle.io/owner-namespace: fleet-default
  creationTimestamp: "2022-02-16T06:45:12Z"
  labels:
    objectset.rio.cattle.io/hash: 0621cab6176495a2ed1be99c2047fb160f682a0f
  name: rke2-in-jeos-pool1-36980d8b-c7dbw-machine-state
  namespace: fleet-default
  resourceVersion: "72295"
  uid: 27fa79e5-dab8-4fcc-8297-48801ac654b3
type: rke.cattle.io/machine-state
```
* HarvesterMachine
```
apiVersion: rke-machine.cattle.io/v1
kind: HarvesterMachine
metadata:
  annotations:
    cluster.x-k8s.io/cloned-from-groupkind: HarvesterMachineTemplate.rke-machine.cattle.io
    cluster.x-k8s.io/cloned-from-name: rke2-in-jeos-pool1-36980d8b
  creationTimestamp: "2022-02-16T06:45:10Z"
  finalizers:
  - wrangler.cattle.io/machine-provision-remove
  generation: 2
  labels:
    cluster.x-k8s.io/cluster-name: rke2-in-jeos
    cluster.x-k8s.io/control-plane: "true"
    cluster.x-k8s.io/deployment-name: rke2-in-jeos-pool1
    machine-template-hash: "3144125949"
    rke.cattle.io/control-plane-role: "true"
    rke.cattle.io/etcd-role: "true"
    rke.cattle.io/worker-role: "true"
  name: rke2-in-jeos-pool1-36980d8b-c7dbw
  namespace: fleet-default
  ownerReferences:
  - apiVersion: cluster.x-k8s.io/v1alpha4
    blockOwnerDeletion: true
    controller: true
    kind: Machine
    name: rke2-in-jeos-pool1-7588569f8f-9c8r6
    uid: e8fcee9c-02e4-407b-9468-70bc63a092c5
  resourceVersion: "73545"
  uid: 5d216655-153c-4c1b-b89e-0d4ee1974913
spec:
  cloudConfig: ""
  clusterId: ""
  clusterType: ""
  common:
    cloudCredentialSecretName: cattle-global-data:cc-d4lgk
  cpuCount: "2"
  diskBus: virtio
  diskSize: "40"
  imageName: default/image-mvm9q
  keyPairName: ""
  kubeconfigContent: ""
  memorySize: "4"
  networkData: ""
  networkModel: virtio
  networkName: default/vlan1
  networkType: dhcp
  providerID: harvester://51caf10c-a5f7-43d0-b97e-2fa5b7baa714
  sshPassword: ""
  sshPort: "22"
  sshPrivateKeyPath: ""
  sshUser: sles
  userData: cGFzc3dvcmQ6IHRlc3QKY2hwYXNzd2Q6CiAgZXhwaXJlOiBmYWxzZQpzc2hfcHdhdXRoOiB0cnVl
  vmNamespace: default
status:
  addresses:
  - address: 192.168.0.52
    type: InternalIP
  - address: rke2-in-jeos-pool1-36980d8b-c7dbw
    type: Hostname
  cloudCredentialSecretName: cattle-global-data:cc-d4lgk
  conditions:
  - message: ""
    reason: ""
    status: "True"
    type: CreateJob
  - message: creating server (HarvesterMachine) in infrastructure provider
    reason: ""
    status: "False"
    type: Ready
  driverHash: e214c5ba38b83febce25863215f887239afee9b4477aa70b4f76695d53378632
  driverUrl: https://releases.rancher.com/harvester-node-driver/v0.3.4/docker-machine-driver-harvester-amd64.tar.gz
  jobName: rke2-in-jeos-pool1-36980d8b-c7dbw-machine-provision
```

**Comment by thedadams** [comment]
@FrankYang0529 Thank you for this. I now understand why the job is getting deleted. I will have a PR up for this tomorrow. 

I believe I also understand why the unauthorized logs were appearing. The service account gets deleted at the same time as the job and there is a grace period of the job deletion. So the service account and the job get deleted at the same time, Kubernetes will wait for the job to complete gracefully until it kills it, during which time the service account is gone and the job can no longer update the secret.

Thank you for helping me identify the issues here. If you can, please open an issue in rancher/rancher, assign it to me, and put it in the v2.6.4 milestone.

**Comment by abonillabeeche** [comment]
Thanks for working on this issue! Please share the rancher issue so we can follow up on it too...

**Comment by FrankYang0529** [comment]
Since upstream PR is merged, I try to validate with `rancher:v2.6-head`. However, I got the following error. It looks like `docker-machine-driver-harvester` is neither a zip file nor a tar file, so there is an error in the rancher-machine job.

```
Downloading driver from https://192.168.18.6/assets/docker-machine-driver-harvester
driver-hash not provided, will not verify after download
ls: cannot access 'docker-machine-driver-*': No such file or directory

cannot open `' (No such file or directory)
driver file does not seem to be in an accepted format
download of driver from https://192.168.18.6/assets/docker-machine-driver-harvester failed
```

**Comment by thedadams** [comment]
@FrankYang0529 The issue seems to be that the driver wasn't downloaded. The error `ls: cannot access 'docker-machine-driver-*': No such file or directory` happens right after the driver is attempted to be downloaded and before we verify that the downloaded driver is executable, a tar file, or zip file.

This seems to be a problem with a previous change where the rancher-machine job couldn't download the driver in an air-gapped environment. I will reopen that issue and update when it is fixed. Apologies for all the difficulties here.

**Comment by thedadams** [comment]
@FrankYang0529 The change for the air-gap download has been merged into Rancher. Can you try this again, please?

Also, I apologize for all the back and forth here. We don't have a way to test these changes. I am trying to get a Harvester setup that we can use to test these changes. Hopefully this will happen soon.

**Comment by FrankYang0529** [comment]
Hi @thedadams, It's okay. I tried again with `v2.6.4-rc4` and still got error like following.

![Screen Shot 2022-02-24 at 1 08 21 PM](https://user-images.githubusercontent.com/4927361/155461906-1e6089f9-4b9b-4a7b-94d7-eeada312fbb8.png)

**Comment by thedadams** [comment]
OK. Thanks. That is a different problem. I will make updates again today.

**Comment by TachunLin** [comment]
Currently the verification is blocked since we could not provision RKE2 cluster, it will stuck in provisioning as the following
![image](https://user-images.githubusercontent.com/29251855/156388552-9adbe04f-dfac-468b-9974-ec55544d6e2d.png)
 
Tracking in harvester issue #1954 related to https://github.com/rancher/rancher/issues/36716

**Comment by TachunLin** [comment]
Checked with @FrankYang0529, currently this issue can be reproduced only on SLES JeOS image `SLES15-SP3-JeOS.x86_64-15.3-OpenStack-Cloud-GM.qcow2`. 

Currently the verification is blocked since we can' provision RKE2 for SLES JeOS on latest rancher `v2.6-head`
Thus we create a separate issue #1973 to track the follow up fix.

**Comment by TachunLin** [comment]
Verified **fixed** on master-088a020f-head. (03/06) with Rancher `v2.6-head`. (`33db2f5`). Close this issue.

 
### Result
* We can scale pool up by clicking + to add node to RKE2 cluster by node driver
![image](https://user-images.githubusercontent.com/29251855/157243631-19d64790-f726-41ca-83f9-80740852035c.png)
![image](https://user-images.githubusercontent.com/29251855/157244484-05552ad7-9b3d-497c-9b97-2c6e9de1b5ce.png)
![image](https://user-images.githubusercontent.com/29251855/157244555-92ab6fdb-2f34-402f-9570-1659006f0bf1.png)

* Can also click - to scale pool down by deleting a node from existing cluster 
![image](https://user-images.githubusercontent.com/29251855/157249545-1a264946-4651-4643-b182-479f6574588d.png)


### Test Information
* Test Environment: 1 node1 harvester on local kvm machine
* Harvester version: master-088a020f-head. (03/06)
* Rancher version: v2.6-head. (33db2f5)
* Kubernetes version: `v1.21.10+rke2r1`
* OS image: `SLES15-SP3-JeOS.x86_64-15.3-OpenStack-Cloud-GM.qcow2`

### Verify Steps
1. Upload `SLES15-SP3-JeOS.x86_64-15.3-OpenStack-Cloud-GM.qcow2` image to harvester cluster
2. Import harvester in rancher
3. Create cloud credential
4. Provision a RKE2 cluster 
5. Select RKE2 version: `v1.21.10+rke2r1`
6. Add the following cloud init in advance settings according to the workaround 
https://github.com/harvester/harvester/issues/1973#issuecomment-1061604853 
  ```
  #cloud-config
  runcmd:
  - zypper ar -cfG 'https://mirrors.tuna.tsinghua.edu.cn/opensuse/tumbleweed/repo/oss/' tuna-oss
  - zypper ar -cfG 'https://mirrors.tuna.tsinghua.edu.cn/opensuse/tumbleweed/repo/non-oss/' tuna-non-oss
  - zypper ref
  - zypper -n in glibc-2.35-2.1.x86_64 glibc-locale-base-2.35-2.1.x86_64 filesystem-84.87-3.2.x86_64 apparmor-parser-3.0.4-1.2.x86_64
  ```
7. Click + to scale pool up by adding node after provisioning RKE2 complete,
  ![image](https://user-images.githubusercontent.com/29251855/157246863-d31517eb-df23-4962-8e60-3cd97d50d5dc.png)

8. Check can add node correctly to existing rke2 cluster 
9. Click - to scale pool down by deleting a node from existing cluster
